### PR TITLE
LLT-4013: Add conntrack logic for ICMP

### PR DESCRIPTION
**Pull request by mathiaspeters**
### Description
So far there has been no conntrack logic for ICMP. The firewall flow for an inbound ICMP packet has been:
1. If peer is on whitelist, allow
2. If ICMP packet type is blocked, block
3. Allow

With these changes, outbound ICMP connections create cache entries, and the firewall flow becomes:
1. If peer is on whitelist, allow
2. If peer connection is in the cache, allow
3. If ICMP packet type is blocked, block
4. Allow

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean
- [x] README.md is updated
- [x] changelog.md is updated

### CODE REVIEW

**Review by jjanowsk — COMMENTED** [review]
+1

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:345]: Fixed

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: Good suggestion! Added

**Review by jjanowsk — APPROVED** [review]

**Comment by lcruz99** [comment]
Do we expect to accept inbound ICMP packet even when `allow_incoming_connections=False`?

**Comment by mathiaspeters** [comment]
@lcruz99 @jjanowsk Can you have another look at this?

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:667]: I don't think it makes sense now. With the previous logic we allowed all incoming traffic, so basically we've been opening the firewall for all ICMP traffic. If the logic is that we should only allow one response packet, then it only  makes sense to let the packet that really is the response, not any packet. So the only candidate for me that should be placed in the cache is an echo req, and it should be removed by a matching echo response. That's it. With this implementation by sending echo req we allow the other side to send an echo req to us which seems strange.

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:289]: I think identifier should be u32. ICMP identifier + seq number. Now if one ping process sends two ping requests they will have the same identifier but different seq number. In this implementation the second response will not get through, because there will be only one entry in the cache which will get removed by the first response.

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:673]: And maybe we want to think about letting BLOCKED_TYPES through in case of whitelisted peers? Not sure about that one. Just opening a discussion

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:662]: Also this should be filled in this function as well in case of echo response I guess

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:673]: Never mind. I can see now that the whitelisting by peer is done earlier before this function

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:667]: Ok more candidates for the cache: Basically all from BLOCKED_TYPES and ICMPV6_BLOCKED_TYPES, but only those.

Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:667]: And also because now we have more packet types in the cache, not only echo, the ICMP type and maybe even code, should also be part of the cache key.

Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:667]: And one additional note: The key in cache should probably have repose type. So the logic would go: when we have outbound echo request we put the key in cache with the same code, identifier and seq number and with echo response type. And when this response comes we remove this key. That's it. I've also  checked that all ipv4 req/resp packets that we have in BLOCKED_TYPES have identifier and seq number

File: crates/telio-firewall/src/firewall.rs
Comment (lcruz99) [inline crates/telio-firewall/src/firewall.rs:695]: lint nagger is complaining about the `unwrap`

**Review by lcruz99 — COMMENTED** [review]
+1.0

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:691]: There is no test for this, if matching replies to requests work. You should add it. But I believe it does not work. Request and replies should produce the same key in the cache. Here outgoing echo request produces different key then incoming echo response, unless I'm missing something. I think it should be sth like:
```
            0 => (0, true),     // echo reply message
            8 => (0, true),     // echo request message
            13 => (14, true),   // timestamp message
            14 => (14, true),   // timestamp reply message
            ...
```
Additionally incoming requests should be just blocked without looking into the cache.

Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:691]: But now thinking about it, we only have requests on `ICMP_BLOCKED_TYPES` lists, so actually the logic that we wanted is already implemented on main. The requests do not go through unless the peer is whitelisted and the responses do come through. What am I missing? :D

Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:691]: You're right about the first comment here, I will update it and add a test.

Regarding the second comment here, the logic on main is: 
1. If whitelisted, allow
2. If on block list, reject
3. allow
and with the changes in this PR, the logic is:
1. if whitelisted, allow
2. if connection is a direct response, allow
3. if on block list, reject
4. allow

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: I updated a function to create this key, that handles this and makes sure it's populated when it should be

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: Fixed

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: Fixed

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: All good points, and all addressed

**Comment by mathiaspeters** [comment]
@lcruz99 @jjanowsk I've addressed your comments, please have another look

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:695]: TBH this seems strange to me :) Now the only added value I see is that we do not allow ICMP responses if the request was not initiated by this peer. But we allow all other ICMP packets in anyway, so I'm not sure what's so special about responses that we want to block them, but not block other icmp packets that are not req/resp. Also now the implementation contains a bug. I've hinted this in my first msg by: "Additionally incoming requests should be just blocked without looking into the cache." Because now the remote peer can send echo request to us when we send the request to him, and that for sure should be not allowed if the peer is not whitelisted.

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: Fixed

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:690]: This should be fixed now. As before, all outbound packets are allowed, and for inbound packets, the logic is now:
1. If whitelisted, allow
2. If on block list, reject (the block list contains the request types we put in the cache, but not the reply types, so we block `EchoRequest` for example)
3. If direct response, allow
4. reject

**Comment by mathiaspeters** [comment]
@lcruz99 @jjanowsk Hopefully my last ping for review here :smile: 
@Jauler please have a look as well

**Comment by jjanowsk** [comment]
So again I have a general question to this :) Do we really need it at this point. We've come to the working implementation which looks and works ok IMO, but the only difference to the `main` is that we filter out echo responses and advert messages. Is it worth doing? Is there any harm in letting in an ICMP echo response? I'll leave some nitpick comments anyway in case we go for it :)

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:673]: Could those be replaced by constants from `IcmpTypes` instead of relying on comments?

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:692]: I think you've again forgot about the check for the length of the payload?

**Comment by mathiaspeters** [comment]
> So again I have a general question to this :) Do we really need it at this point. We've come to the working implementation which looks and works ok IMO, but the only difference to the `main` is that we filter out echo responses and advert messages. Is it worth doing? Is there any harm in letting in an ICMP echo response? I'll leave some nitpick comments anyway in case we go for it :)

With the current solution we're filtering out anything that isn't a direct response to the message types `BLOCKED_TYPES` (4 types for v4 and 4 types for v6), so there is a lot more filtering now than there was before. You can see if the function `build_icmp_key` which ICMP types we create cache keys for, any other type we reject. On top of that, since checking `BLOCKED_TYPES` now happens before checking the cache, `handle_inbound_icmp` only allows the few reply types that get cache keys.

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:692]: `try_into` returns a `Result` with an `Err` if the body is not long enough, which then gets rejected by the if-else below this line

Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:692]: It does not protect you from the panic. `try_into` in this example converts from slice to array. It converts to 4 element array because that's what expected by `u32::from_ne_bytes` and it will fail if you give it a slice of the wrong size. E.g. `icmp_packet.payload()[0..2].try_into()` will return None because the slice is too short. But it does nothing to check that your slice itself is correct, and in that case it might not be correct. So if your slice points to unallocated memory this will panic.

Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:692]: Here's the playground to verify: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=e8bcc5f5b43d0e32c7f62afb6d8522c6

Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:692]: I didn't consider the slicing operation, only the conversion. You are of course correct, and I will add a check

**Comment by jjanowsk** [comment]
> > So again I have a general question to this :) Do we really need it at this point. We've come to the working implementation which looks and works ok IMO, but the only difference to the `main` is that we filter out echo responses and advert messages. Is it worth doing? Is there any harm in letting in an ICMP echo response? I'll leave some nitpick comments anyway in case we go for it :)
> 
> With the current solution we're filtering out anything that isn't a direct response to the message types `BLOCKED_TYPES` (4 types for v4 and 4 types for v6), so there is a lot more filtering now than there was before. You can see if the function `build_icmp_key` which ICMP types we create cache keys for, any other type we reject. On top of that, since checking `BLOCKED_TYPES` now happens before checking the cache, `handle_inbound_icmp` only allows the few reply types that get cache keys.

You're right, I've missed that. Previous logic was that if the function creating the key returned None the packet was let through.

**Comment by Jauler** [comment]
> > > So again I have a general question to this :) Do we really need it at this point. We've come to the working implementation which looks and works ok IMO, but the only difference to the `main` is that we filter out echo responses and advert messages. Is it worth doing? Is there any harm in letting in an ICMP echo response? I'll leave some nitpick comments anyway in case we go for it :)
> > 
> > 
> > With the current solution we're filtering out anything that isn't a direct response to the message types `BLOCKED_TYPES` (4 types for v4 and 4 types for v6), so there is a lot more filtering now than there was before. You can see if the function `build_icmp_key` which ICMP types we create cache keys for, any other type we reject. On top of that, since checking `BLOCKED_TYPES` now happens before checking the cache, `handle_inbound_icmp` only allows the few reply types that get cache keys.
> 
> You're right, I've missed that. The logic of the previous version was that if the function creating the key returned None the packet was let through. But now I'm worried that by blocking all icmp traffic we can do more harm then good. E.g. Path MTU Discovery procedure can be broken because of blocking "Fragmentation required" message. There are probably more examples like this. So in my opinion blocking just blacklisted packet types like we do on main should be enough. But I would like some additional pair of eyes to review this and participate in this discussion. Sorry for dragging this review ;)

That is a good point. IMHO, the "best" way to filter ICMP error packets, would be according to the IP header stored in the payload of the message (e.g. Normally it starts 8 bytes into the ICMP message, but please double-check this on every message). At first I personally I somehow thought that properly supporting ICMP messages is somewhat low on the priority list. But now that you mention MTU discovery, I am kinda thinking, that it would make sense to be on the safe side and add support for them.

@jjanowsk, would you agree with the approach of taking the IP header of the original message from ICMP error packet and matching it against either TCP or UDP or ICMP [caches](https://github.com/NordSecurity/libtelio/blob/main/crates/telio-firewall/src/firewall.rs#L218)?

File: crates/telio-firewall/src/firewall.rs
Comment (Jauler) [inline crates/telio-firewall/src/firewall.rs:390]: If I remember correctly, Next Header protocol for IPv6 ICMP is different from IPv4, so we should have a case for `IpNextHeaderProtocols::Icmpv6` (refernece: https://docs.rs/pnet/latest/pnet/packet/ip/IpNextHeaderProtocols/constant.Icmpv6.html)

File: crates/telio-firewall/src/firewall.rs
Comment (Jauler) [inline crates/telio-firewall/src/firewall.rs:674]: :thinking: In this case I believe we do not expect replies when an outbound "reply" packet is sent, therefore we probably just need to populate cache only if outbound request packet passes.

Or are there any cases which I am missing?

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:390]: In Matas' PR he handles this case, but it's essentially just generics and calling the same function

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:674]: Would we ever send an outbound reply? I can add a special case for that if you think it's relevant

**Comment by mathiaspeters** [comment]
@jjanowsk There is a follow-up ticket to add handling of ICMP error types that I will start with as soon as this is okay to merge. So disregarding that, is there anything left here that is a blocker or needs fixing in your opinion?

File: crates/telio-firewall/src/firewall.rs
Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:674]: Good point. I was thinking that we don't have outbound replies but of course we do. I will add special handling of that

**Comment by jjanowsk** [comment]
> @jjanowsk There is a follow-up ticket to add handling of ICMP error types that I will start with as soon as this is okay to merge. So disregarding that, is there anything left here that is a blocker or needs fixing in your opinion?

Yes, my question stays: What do we gain by complicating ICMP rules so much? Now if we accept this we need to fix error messages by complicating it even further. What is so wrong with the current approach we have on main where we just filter out requests but let all other ICMP traffic through. Maybe it would be better to first think if there are any other ICMP codes that we want to filter out?

**Comment by mathiaspeters** [comment]
> > @jjanowsk There is a follow-up ticket to add handling of ICMP error types that I will start with as soon as this is okay to merge. So disregarding that, is there anything left here that is a blocker or needs fixing in your opinion?
> 
> Yes, my question stays: What do we gain by complicating ICMP rules so much? Now if we accept this we need to fix error messages by complicating it even further. What is so wrong with the current approach we have on main where we just filter out requests but let all other ICMP traffic through. Maybe it would be better to first think if there are any other ICMP codes that we want to filter out?

@Jauler I'll leave this question to you

**Comment by Jauler** [comment]
> > @jjanowsk There is a follow-up ticket to add handling of ICMP error types that I will start with as soon as this is okay to merge. So disregarding that, is there anything left here that is a blocker or needs fixing in your opinion?
> 
> Yes, my question stays: What do we gain by complicating ICMP rules so much? Now if we accept this we need to fix error messages by complicating it even further. What is so wrong with the current approach we have on main where we just filter out requests but let all other ICMP traffic through. Maybe it would be better to first think if there are any other ICMP codes that we want to filter out?

Generally the issue is that with stateless approach we do let "unexpected" replies (and other type packets) through firewall e.g. imagine a situation where node A does not allow incoming connections from node B, all ICMP packets except for request ones will be let through the firewall in this situation. And this "incoming connection" permission does give the expectation that only packets belonging to the communication initiated by the node A will be let through firewall.

While I would not consider this issue to be high priority, with this PR - we already have all the required state tracking implemented, and ICMP error handling does not require additional state tracking, only additional allow condition for inbound packets.

**Comment by mathiaspeters** [comment]
@jjanowsk Thoughts about :point_up:?

**Comment by jjanowsk** [comment]
Sorry, I've missed the reply. As I wrote. I don't have strong opinion on that, I just wanted to start a discussion: is it really worth adding the complexity. If you think it is worth it then I'm fine with that. I'll quicky glance over the changes again now and approve it if I won't find anything

**Comment by jjanowsk** [comment]
Ohh, actually I see from comments that you still want to do some changes, so I'll wait with final review for those. Ping me when it's done

**Comment by mathiaspeters** [comment]
@jjanowsk Please have another look

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:676]: I think you have this logic backwards. You don't really care about replies. You care about not letting in requests. So I would imagine having

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:1518]: this helper is copied 3 times. I think it can be extracted to the separate function

File: crates/telio-firewall/src/firewall.rs
Comment (jjanowsk) [inline crates/telio-firewall/src/firewall.rs:1661]: Why `i == 0 && j == 0`? Advert should also be a valid response for solicit

Comment (jjanowsk) [reply crates/telio-firewall/src/firewall.rs:1661]: I think it should be `i == j` as for v4 messages. If it's not then there is a bug somewhere. It might be that you overwrite some bytes in `make_packet` function for those messages even though they do not have id and seq fields.

Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs:1661]: The complicated thing here, which I haven't really thought about how to approach is that `BLOCKED_TYPES` for IPv6 contains `RouterAdvert`, `NeighborSolicit` and `NeighborAdvert`. Since those types are in `BLOCKED_TYPES` I added them to the cache handling like the other types, but they are essentially unusable given that they're in `BLOCKED_TYPES` so it would probably make more sense to remove them from the cache handling completely. That would reject any inbound ICMP packet with those type, but then again, that is already happening

Comment (mathiaspeters) [reply crates/telio-firewall/src/firewall.rs]: I removed handling of those ICMP types from the cache since they will be blocked anyways, and I simplified this test

**Review by jjanowsk — DISMISSED** [review]
+1

**Review by jjanowsk — APPROVED** [review]

**Pull request merged**
